pacman/operations/placer_algorithms/one_to_one_placer.py

```diff
@@ -35,14 +93,19 @@ def __call__(self, machine_graph, machine):
 
         # Get which vertices must be placed on the same chip as another vertex
         same_chip_vertex_groups = get_same_chip_vertex_groups(machine_graph)
-        sorted_vertices = self._sort_vertices_for_one_to_one_connection(
-            machine_graph, same_chip_vertex_groups)
+
+        # Work out the vertices that should be on the same chip by one-to-one
+        # connectivity
+        one_to_one_groups = group_vertices(
+            machine_graph.vertices, functools.partial(
+                _find_one_to_one_vertices, graph=machine_graph))
 
         return self._do_allocation(
-            sorted_vertices, machine, same_chip_vertex_groups, machine_graph)
+            one_to_one_groups, same_chip_vertex_groups, machine, machine_graph)
```

Comment:
one_to_one_groups and same_chip_vertex_groups hold basically the same thing and have the same format. maybe its worth merging them so that its just one thing handing down to the placer code? so that its easier if there were more constraints. like sdram links

Reply:
this mainly as there's going to be other situations. like SDRAM link checks. maybe a list of functions should be handed down to group_vertices()? so that the common code can just handle them all????

Reply:
i note that the sdram links are already in there. merged into the common code same chip checker. this makes me wonder if they should be separate ones to reflect that there's more...... unknown.

Reply:
The issue is that the one-to-one "constraint" is soft i.e. I don't have to put it on the same chip.  So if the placement of a one-to-one group can't be done, the algorithm has to fall back on trying normal placement.  However this must still respect the SDRAM edges and same-chip-as constraints.

Reply:
but thats reflected by the hard ones being checked by chip constraints later on. thus why you can get away with it

Reply:
note next comment which set me off there

Reply:
I don't think anything else in the code makes this happen though... i.e. there is nothing that checks chip constraints later on (that I know of).

Reply:
119 to 126

Reply:
That only checks for ChipAndCoreConstraint.

Reply:
So this will only place vertices on the same chip if one of those vertices is further constrained.  After that, it tries one-to-one placement of all vertices, or if that fails, it does same-chip-as.

Reply:
right, so the chip and core is a hard chip placement. and is done as we always done it. 

after that it only does 1 by 1 if there's no chip with enough cores. if it can do it, it does it as a group first, then if that fails, then tries with 1 by 1. 

i do agree that there's something i missed, which is with sdram links, if you fail as a group, it'll still do 1 by 1, which is wrong on a sdram link aspect...... but yours has that same issue. which kinda implies the hard constraint check is not enough, but if you got that to cover sdram links as well, you could merge them into 1 dict.

Reply:
SDRAM links are handled in the same_chip_as grouping.  The one-to-one dict includes SDRAM linked and same-chip-as linked vertices i.e. it is already merged in.  See lines 74-76.

Reply:
bugger, it took me a while to see that sneaky bit. But then why do lines 95? why have the second one in the first place? when the data is in first param for _do_allocation??

Reply:
You need the second one in case the first doesn't work - if you can't allocate all the one-to-one vertices on the same chip, you can still get a working placement with the same-chip-as vertices.  The one-to-one "constraint" is a soft one but the same-chip-as is not.

Reply:
Once you have them merged, you can't tell which is which.

Reply:
but my point is that you dont care. i think.

Reply:
and the issue of sdram links being split over multiple chips is still there i think

Reply:
OK, then I am telling you that you are wrong not to!  We want to try to put them all on the same chip, but if that fails, you don't need to fail completely (if you do, you are failing where another placer would have worked and that seems like a bad idea).  And if you don't fail completely, you need to make sure the actual constraints requested are respected.

Reply:
SDRAM links are being respected by the "allocate_individual" function.  If you can write a test that says they are not though, I would be willing to fix it.

Reply:
but your not going to fail when its the one to one constrained, youll try first on all on same chip. as line 156 ensures it wont try individual first, unless it doesnt fit on a chip, which fine, its individual. so assuming fits on chip,  itll go to line 166, which goes to 182 which does group then if that fails, (170) itll do indiviudal again. But you will not fail when your meant to when dealing with sdram links. as sdram links have no placement constraint associated with them before placer engages and so will be in weak linked, go through the same lines, at which point line 170 logic is wrong. as a sdram link work over multiple chips.

now given your so strongly in beleif im wrong, what have i missed there?

Reply:
The reason that these are not done in the same context as the ChipAndCore constraint is that they don't have a specific place to go.  So long as there is at least one chip with enough space, they can go on that chip.

Reply:
ya win. but ive left the test in there. least itll stop my logic path in future. 

but for my defence. line 201 " _allocate_individual(" does not do what it says. it does allocate_group. so its a bit misleading, which could be part of what took me down that road. 

and ill admit if you replace same_chip_as with one to one, it blows up, but id like to figure out why.  as i dont really get why it doesnt work. when it contains the same data in there.

Reply:
OK, to help with understanding, the sets in the one-to-one dict are super sets of those in the same-chip-as dict.  Say you have a vertex A which is one-to-one connected to vertex B which has to go on the same chip as vertex C, and one chip left with only 2 cores on it, and one with only 1 core e.g.:
A --> B <--> C  
Chip 1: 2 cores
Chip 2: 1 core

This means that the one-to-one dict has:
A: [A, B, C]
B: [A, B, C]
C: [A, B, C]

and the same-chip-as dict has:
A: [A]
B: [B, C]
C: [B, C]

Say now you try to place B, and it tries to place all three, but there isn't a chip with 3 cores, so that fails.  So you say to place vertex B "individually", which results in B and C being placed.

Does that help?